# Sold Data EDA (Week1) 
Author : Yoora Choi

In [19]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [20]:
# Import listed data (2024.01-2026.03)
df = pd.read_csv("CRMLSSold_merge.csv")

/var/folders/d2/04b97b6n77b6vr8qn0rd3q340000gn/T/ipykernel_54740/641313067.py:2: DtypeWarning: Columns (0,1,4,9,78,79,80,82,83) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("CRMLSSold_merge.csv")


## 1.Data overview

In [21]:
# Check data structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 591104 entries, 0 to 591103
Data columns (total 84 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   BuyerAgentAOR                 519595 non-null  object 
 1   ListAgentAOR                  523058 non-null  object 
 2   Flooring                      347460 non-null  object 
 3   ViewYN                        531914 non-null  object 
 4   WaterfrontYN                  322 non-null     object 
 5   BasementYN                    9738 non-null    object 
 6   PoolPrivateYN                 516968 non-null  object 
 7   OriginalListPrice             589393 non-null  float64
 8   ListingKey                    591104 non-null  int64  
 9   ListAgentEmail                547714 non-null  object 
 10  CloseDate                     591104 non-null  object 
 11  ClosePrice                    591097 non-null  float64
 12  ListAgentFirstName            587671 non-nul

In [22]:
# check the data structure
df.head(5)

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,ListAgentFirstName,ListAgentLastName,Latitude,Longitude,UnparsedAddress,PropertyType,LivingArea,ListPrice,DaysOnMarket,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,FireplacesTotal,AssociationFeeFrequency,AboveGradeFinishedArea,ListingKeyNumeric,MLSAreaMajor,TaxAnnualAmount,CountyOrParish,MlsStatus,ElementarySchool,AttachedGarageYN,ParkingTotal,BuilderName,PropertySubType,LotSizeAcres,SubdivisionName,BuyerOfficeAOR,YearBuilt,StreetNumberNumeric,ListingId,BathroomsTotalInteger,City,TaxYear,BuildingAreaTotal,BedroomsTotal,ContractStatusChangeDate,ElementarySchoolDistrict,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BelowGradeFinishedArea,BusinessType,StateOrProvince,CoveredSpaces,MiddleOrJuniorSchool,FireplaceYN,Stories,HighSchool,Levels,LotSizeDimensions,LotSizeArea,MainLevelBedrooms,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,BuyerAgencyCompensation,latfilled,lonfilled
0,Mlslistings,Mlslistings,"Carpet,Tile,Wood",True,NaN,NaN,False,499000.0,551985747,jwachter@cbnorcal.com,2024-01-26,240000.0,Joan,Wachter,NaN,NaN,1 Baldwin Avenue 411,Residential,1140.0,295000.0,777,Coldwell Banker Realty,Bay Area Senior Services,NaN,Joan Wachter,NaN,NaN,ML5090968,Scott,Withrow,NaN,Monthly,NaN,551985747,699 - Not Defined,NaN,San Mateo,Closed,NaN,False,1.0,NaN,Condominium,NaN,NaN,Mlslistings,1988.0,1.0,ML81865679,2.0,San Mateo,NaN,NaN,2.0,2024-01-26,NaN,NaN,2023-11-22,2021-10-06,NaN,NaN,CA,NaN,NaN,True,NaN,NaN,NaN,NaN,NaN,NaN,False,1.0,Other,94401,6472.0,NaN,NaN,CRMLS,CRMLS_MLSL,NaN,NaN,NaN,NaN
1,HighDesert,HighDesert,NaN,NaN,NaN,NaN,NaN,0.0,535486633,eabrown@lee-associates.com,2024-01-24,950.0,Elizabeth,Brown,34.529407,-117.319674,15770 Mojave Drive L,CommercialLease,NaN,808.0,901,Lee & Associates,Lee & Associates,NaN,Elizabeth Brown,NaN,NaN,HD11042,Elizabeth,Brown,NaN,NaN,NaN,535486633,NaN,NaN,San Bernardino,Closed,NaN,NaN,NaN,NaN,Retail,1.2000,NaN,HighDesert,1980.0,15770.0,538013,NaN,Victorville,NaN,NaN,NaN,2024-01-24,NaN,NaN,2024-01-24,2021-08-06,NaN,Other,CA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,52320.0,NaN,NaN,NaN,NaN,92394,NaN,52320.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN
2,OrangeCounty,OrangeCounty,NaN,True,NaN,NaN,NaN,75000.0,529986282,Joe@9WINWIN.com,2024-01-16,45000.0,Joseph,Piteleski,35.618009,-118.473141,0 Shadow mountain Dr.,Land,NaN,75000.0,865,Keller Williams OC Coastal Realty,Keller Williams OC Coastal Realty,NaN,Joseph Piteleski,NaN,NaN,spitejos,Joseph,Piteleski,NaN,NaN,NaN,529986282,LKIS - Lake Isabella,NaN,Kern,Closed,NaN,NaN,NaN,NaN,NaN,4.9900,NaN,OrangeCounty,NaN,0.0,OC21157144,NaN,Lake Isabella,NaN,NaN,NaN,2024-01-16,NaN,NaN,2023-12-22,2021-07-18,NaN,NaN,CA,NaN,NaN,NaN,NaN,NaN,NaN,NaN,217364.0,NaN,False,NaN,NaN,93240,NaN,217364.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN
3,InlandValleys,InlandValleys,NaN,True,NaN,NaN,NaN,199000.0,529618166,carolthefinder@yahoo.com,2024-01-08,141500.0,CAROL,SMITH,34.396733,-117.220954,20171 Riverview Road,Land,NaN,145000.0,830,Keller Williams Realty Phelan,Coldwell Banker Home Source,Keller Williams Realty Phelan,CAROL SMITH,DINA,VELEZ,HD12177,Stacey,Hart,NaN,NaN,NaN,529618166,NaN,NaN,San Bernardino,Closed,NaN,NaN,NaN,NaN,NaN,5.0000,NaN,HighDesert,NaN,20171.0,537165,NaN,Apple Valley,NaN,NaN,NaN,2024-01-08,NaN,NaN,2023-10-24,2021-07-14,NaN,NaN,CA,NaN,NaN,NaN,NaN,NaN,NaN,Irregular,217800.0,NaN,False,NaN,NaN,92308,NaN,217800.0,NaN,CRMLS,CRMLS_CRM,NaN,NaN,NaN,NaN
4,SouthwestRiversideCounty,SouthwestRiversideCounty,NaN,True,NaN,NaN,NaN,19500.0,522614340,jtavisola@tavisola.com,2024-01-17,15000.0,Jeremie,Tavisola,34.454756,-117.782248,0 Vac/Vic Avenue Z/195 Ste,Land,NaN,19500.0,805,Realty 

In [23]:
# Summary statistics for numerical variable
df.describe()

,OriginalListPrice,ListingKey,ClosePrice,Latitude,Longitude,LivingArea,ListPrice,DaysOnMarket,FireplacesTotal,AboveGradeFinishedArea,ListingKeyNumeric,TaxAnnualAmount,ParkingTotal,LotSizeAcres,YearBuilt,StreetNumberNumeric,BathroomsTotalInteger,TaxYear,BuildingAreaTotal,BedroomsTotal,ElementarySchoolDistrict,BelowGradeFinishedArea,CoveredSpaces,Stories,LotSizeArea,MainLevelBedrooms,GarageSpaces,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,BuyerAgencyCompensation
count,5.893930e+05,5.911040e+05,5.910970e+05,572088.000000,572260.000000,5.493810e+05,5.901940e+05,591104.000000,0.0,0.0,5.911040e+05,2.507000e+03,5.724340e+05,5.351400e+05,566191.000000,5.896420e+05,563705.000000,367.000000,76569.000000,551747.000000,0.0,2590.000000,0.0,472780.000000,5.370080e+05,313623.000000,515436.000000,407872.000000,5.365510e+05,0.0,67780.000000
mean,9.061054e+05,1.095900e+09,8.745520e+05,34.527336,-118.451203,3.497140e+03,8.399954e+05,43.284468,NaN,NaN,1.095900e+09,1.847250e+04,7.608553e+00,7.635590e+02,1978.099500,2.021475e+04,2.445955,2022.970027,2552.308327,3.080512,NaN,106.557996,NaN,1.364916,4.816410e+04,1.931140,1.765978,185.364682,4.701315e+05,NaN,194.406143
std,5.745218e+06,3.325829e+07,5.275359e+06,1.614965,3.113594,1.226721e+06,1.275633e+06,69.834146,NaN,NaN,3.325829e+07,2.105646e+05,3.439081e+03,5.063346e+05,27.086022,1.468013e+06,1.417313,2.429706,5192.219818,1.399860,NaN,799.825686,NaN,0.481407,2.238151e+06,6.252108,3.877713,1633.587365,2.078122e+07,NaN,2152.716451
min,0.000000e+00,4.214490e+08,0.000000e+00,-117.472493,-177.646696,0.000000e+00,0.000000e+00,-288.000000,NaN,NaN,4.214490e+08,0.000000e+00,-1.430000e+02,0.000000e+00,1776.000000,0.000000e+00,0.000000,2009.000000,0.000000,0.000000,NaN,0.000000,NaN,1.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000e+00,NaN,0.000000
25%,6.500000e+04,1.075085e+09,5.250000e+04,33.739574,-118.538038,1.183000e+03,6.250000e+04,9.000000,NaN,NaN,1.075085e+09,2.445000e+02,2.000000e+00,1.194000e-01,1960.000000,8.320000e+02,2.000000,2023.000000,1161.000000,2.000000,NaN,0.000000,NaN,1.000000,4.941000e+03,1.000000,1.000000,0.000000,5.185500e+03,NaN,2.000000
50%,6.399900e+05,1.095288e+09,6.299000e+05,34.038491,-118.043184,1.579000e+03,6.299000e+05,22.000000,NaN,NaN,1.095288e+09,6.330000e+03,2.000000e+00,1.677000e-01,1979.000000,2.902000e+03,2.000000,2023.000000,1699.000000,3.000000,NaN,0.000000,NaN,1.000000,7.025000e+03,2.000000,2.000000,0.000000,7.292000e+03,NaN,2.500000
75%,1.089000e+06,1.117367e+09,1.070000e+06,34.304252,-117.329031,2.150000e+03,1.050000e+06,54.000000,NaN,NaN,1.117367e+09,1.942500e+04,3.000000e+00,2.985000e-01,1999.000000,1.110500e+04,3.000000,2024.000000,2651.000000,4.000000,NaN,0.000000,NaN,2.000000,1.148000e+04,3.000000,2.000000,295.000000,1.295800e+04,NaN,2.500000
max,1.390000e+09,1.156473e+09,9.895000e+08,157.000000,329.000000,9.090909e+08,1.375000e+08,12430.000000,NaN,NaN,1.156473e+09,1.025400e+07,2.593669e+06,3.702600e+08,2026.000000,7.212710e+08,175.000000,2026.000000,462607.000000,123.000000,NaN,17424.000000,NaN,2.000000,9.187423e+08,3333.000000,600.000000,750000.000000,5.193920e+09,NaN,100000.000000


In [24]:
# Summary statistics for non-numerical variable
df.describe(include = 'object')

,BuyerAgentAOR,ListAgentAOR,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,ListAgentEmail,CloseDate,ListAgentFirstName,ListAgentLastName,UnparsedAddress,PropertyType,ListOfficeName,BuyerOfficeName,CoListOfficeName,ListAgentFullName,CoListAgentFirstName,CoListAgentLastName,BuyerAgentMlsId,BuyerAgentFirstName,BuyerAgentLastName,AssociationFeeFrequency,MLSAreaMajor,CountyOrParish,MlsStatus,ElementarySchool,AttachedGarageYN,BuilderName,PropertySubType,SubdivisionName,BuyerOfficeAOR,ListingId,City,ContractStatusChangeDate,CoBuyerAgentFirstName,PurchaseContractDate,ListingContractDate,BusinessType,StateOrProvince,MiddleOrJuniorSchool,FireplaceYN,HighSchool,Levels,LotSizeDimensions,NewConstructionYN,HighSchoolDistrict,PostalCode,OriginatingSystemName,OriginatingSystemSubName,BuyerAgencyCompensationType,latfilled,lonfilled
count,519595,523058,347460,531914,322,9738,516968,547714,591104,587671,591043,590204,591104,591104,583783,126899,590943,122194,122446,589274,588480,590768,174710,526103,591104,591104,65586,427245,22579,545784,208543,565619,591104,590664,590392,47999,579280,591021,1632,591104,66027,539436,86048,509828,31011,520943,393965,590939,59998,59998,67747,95793,95793
unique,63,59,342,2,1,1,2,92285,821,17202,38174,551830,8,21491,24107,6206,87918,7347,15393,134761,20890,43676,4,1126,63,1,1419,2,3935,34,19666,69,590530,1237,821,4841,1237,1816,760,25,669,2,525,19,13078,2,446,4145,1,10,3,2,2
top,OrangeCounty,OrangeCounty,Wood,True,True,True,False,homes@opendoor.com,2024-06-28,David,Lee,1225 Vienna Drive,Residential,Compass,Compass,Compass,Shaheen Moarefi,Michael,Myatt,NONMEMBER,NONE,NONE,Monthly,699 - Not Defined,Los Angeles,Closed,Other,True,Lennar,SingleFamilyResidence,Other,OrangeCounty,250040515SD,Los Angeles,2024-06-28,Michael,2024-07-01,2025-05-01,Restaurant,CA,Other,True,Other,One,50x135,False,Los Angeles Unified,92618,CRMLS,CRMLS_CRM,Item1,False,False
freq,50979,51438,39075,323372,322,9738,456461,1068,1897,7357,3527,54,397210,37499,35539,13492,998,1890,891,10868,9955,9859,163800,54337,179071,591104,10219,332263,2001,360957,5872,53287,4,39599,1898,689,1287,1656,166,591027,7691,331481,3017,297367,584,500092,37380,3460,59998,34276,61533,50537,50639


## 2. Missing data check(null)

In [25]:
# null count
null_count = df.isnull().sum()

# null percentage (%)
null_percent = (df.isnull().sum() / len(df)) * 100

# data frame
null_summary = pd.DataFrame({
    'null_count': null_count,
    'null_percent': null_percent
}).sort_values(by='null_percent', ascending=False)

null_summary['null_percent'] = null_summary['null_percent'].round(2)
null_summary

,null_count,null_percent
AboveGradeFinishedArea,591104,100.00
FireplacesTotal,591104,100.00
ElementarySchoolDistrict,591104,100.00
CoveredSpaces,591104,100.00
MiddleOrJuniorSchoolDistrict,591104,100.00
WaterfrontYN,590782,99.95
TaxYear,590737,99.94
BusinessType,589472,99.72
TaxAnnualAmount,588597,99.58
BelowGradeFinishedArea,588514,99.56


## 3. Check duplicates

### 3.1 duplicated columns

In [26]:
# check duplicate columns
pairs = []

for col in df.columns:
    if '.' in col:
        base_col = col.split('.')[0]
        if base_col in df.columns:
            pairs.append((base_col, col))

pairs

[]

### 3.2 duplicated rows

In [27]:
# check completely duplicated rows
#df[df.duplicated(keep=False)]

In [28]:
df.duplicated().sum()

np.int64(140)

# Data Analysis Summary (CRMLSSold data (2024.01-2026.03))

## 1. Basic Information
- **Total Rows:** 591,104
- **Total Columns:** 84
- **Data Period:** January 2024 – March 2026

---

## 2. Column Structure
- **Data Types:** float64(28), int64(3), object(53)

---

## 3. Missing Data

### Completely Empty Columns (100% Null) → Consider Dropping
- `AboveGradeFinishedArea`, `FireplacesTotal`, `ElementarySchoolDistrict`, `CoveredSpaces`, `MiddleOrJuniorSchoolDistrict`

### High Missing Rate (80–100%)
- `WaterfrontYN` (99.95%), `TaxYear` (99.94%), `BusinessType` (99.72%), `TaxAnnualAmount` (99.58%), `BelowGradeFinishedArea` (99.56%), `BasementYN` (98.35%), `BuilderName` (96.18%), `LotSizeDimensions` (94.75%), `CoBuyerAgentFirstName` (91.88%), `OriginatingSystemName` (89.85%), `OriginatingSystemSubName` (89.85%), `ElementarySchool` (88.90%), `MiddleOrJuniorSchool` (88.83%), `BuyerAgencyCompensationType` (88.54%), `BuyerAgencyCompensation` (88.53%), `BuildingAreaTotal` (87.05%), `HighSchool` (85.44%), `latfilled` (83.79%), `lonfilled` (83.79%)

### High Missing Rate (70–80%)
- `CoListAgentFirstName` (79.33%), `CoListAgentLastName` (79.29%), `CoListOfficeName` (78.53%)

### Moderate Missing Rate (10–70%)
- `AssociationFeeFrequency` (70.44%), `SubdivisionName` (64.72%), `MainLevelBedrooms` (46.94%), `Flooring` (41.22%), `HighSchoolDistrict` (33.35%), `AssociationFee` (31.00%), `AttachedGarageYN` (27.72%), `Stories` (20.02%), `Levels` (13.75%), `GarageSpaces` (12.80%), `PoolPrivateYN` (12.54%), `BuyerAgentAOR` (12.10%), `NewConstructionYN` (11.87%), `ListAgentAOR` (11.51%), `MLSAreaMajor` (11.00%), `ViewYN` (10.01%)

### Low Missing Rate (0–10%)
- `LotSizeAcres` (9.47%), `LotSizeSquareFeet` (9.23%), `LotSizeArea` (9.15%), `FireplaceYN` (8.74%), `PropertySubType` (7.67%), `ListAgentEmail` (7.34%), `LivingArea` (7.06%), `BedroomsTotal` (6.66%), `BathroomsTotalInteger` (4.64%), `BuyerOfficeAOR` (4.31%), `YearBuilt` (4.21%), `Latitude` (3.22%), `Longitude` (3.19%), `ParkingTotal` (3.16%), `PurchaseContractDate` (2.00%), `BuyerOfficeName` (1.24%), `ListAgentFirstName` (0.58%), `BuyerAgentFirstName` (0.44%), `BuyerAgentMlsId` (0.31%), `OriginalListPrice` (0.29%), `StreetNumberNumeric` (0.25%), `ListPrice` (0.15%), `UnparsedAddress` (0.15%), `ContractStatusChangeDate` (0.12%), `City` (0.07%), `BuyerAgentLastName` (0.06%), `PostalCode` (0.03%), `ListAgentFullName` (0.03%), `ListingContractDate` (0.01%), `ListAgentLastName` (0.01%), `ClosePrice` (0.00%)

### No Missing Values
- `ListingKey`, `CloseDate`, `MlsStatus`, `PropertyType`, `DaysOnMarket`, `ListOfficeName`, `ListingKeyNumeric`, `ListingId`, `CountyOrParish`, `StateOrProvince`

---

## 4. Key MLS Fields – Statistical Summary


In [29]:
# Key MLS Fields - Numerical summary
key_num_cols = ['ListPrice', 'ClosePrice', 'LivingArea', 'BedroomsTotal', 'BathroomsTotalInteger', 'DaysOnMarket']
df[key_num_cols].describe().round(2)

,ListPrice,ClosePrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket
count,5.901940e+05,5.910970e+05,5.493810e+05,551747.00,563705.00,591104.00
mean,8.399954e+05,8.745520e+05,3.497140e+03,3.08,2.45,43.28
std,1.275633e+06,5.275359e+06,1.226721e+06,1.40,1.42,69.83
min,0.000000e+00,0.000000e+00,0.000000e+00,0.00,0.00,-288.00
25%,6.250000e+04,5.250000e+04,1.183000e+03,2.00,2.00,9.00
50%,6.299000e+05,6.299000e+05,1.579000e+03,3.00,2.00,22.00
75%,1.050000e+06,1.070000e+06,2.150000e+03,4.00,3.00,54.00
max,1.375000e+08,9.895000e+08,9.090909e+08,123.00,175.00,12430.00


### Data Quality Issues Detected
- **ListPrice / ClosePrice / LivingArea:** Min = 0 and extremely high max values suggest outliers or data entry errors, which requires cleaning.
- **DaysOnMarket:** Negative min value (-288) is invalid, which requires investigation
- **BathroomsTotalInteger:** Max = 175 is likely an error

In [30]:
# Key MLS Fields - Categorical summary
key_cat_cols = ['PropertyType', 'ListOfficeName']
df[key_cat_cols].describe()

,PropertyType,ListOfficeName
count,591104,591104
unique,8,21491
top,Residential,Compass
freq,397210,37499


In [31]:
# Key MLS Fields - Categorical value counts
for col in key_cat_cols:
    print(f"\n--- {col} ---")
    print(df[col].value_counts().head(15))


--- PropertyType ---
PropertyType
Residential            397210
ResidentialLease       135469
Land                    19321
ManufacturedInPark      16058
ResidentialIncome       15831
CommercialSale           3709
CommercialLease          3103
BusinessOpportunity       403
Name: count, dtype: int64

--- ListOfficeName ---
ListOfficeName
Compass                                                  37499
Coldwell Banker Realty                                   25599
Keller Williams Realty                                   10025
First Team Real Estate                                    7780
Berkshire Hathaway HomeServices California Properties     7719
eXp Realty of California Inc                              5940
Real Broker                                               5774
Equity Union                                              4833
Intero Real Estate Services                               4785
eXp Realty of California, Inc.                            4092
The Agency                    


## 5. Duplicate Rows
- 140 rows, which requires cleaning.

---

## 6. Next Steps
1. Drop 100% null columns: `AboveGradeFinishedArea`, `FireplacesTotal`, `ElementarySchoolDistrict`, `CoveredSpaces`, `MiddleOrJuniorSchoolDistrict`
2. Investigate and clean outliers in `ListPrice`, `ClosePrice`, `LivingArea`, and `DaysOnMarket`
3. Fix data types for the following columns:
   - `CloseDate`, `PurchaseContractDate`, `ListingContractDate`, `ContractStatusChangeDate` : object → datetime
   - `YearBuilt`, `BedroomsTotal`, `BathroomsTotalInteger`, `Stories`, `GarageSpaces`, `ParkingTotal`, `MainLevelBedrooms` : float64 → int64 (after handling nulls)
   - `AttachedGarageYN`, `FireplaceYN`, `NewConstructionYN`, `PoolPrivateYN`, `ViewYN`, `BasementYN`, `WaterfrontYN` : object → bool
4. Reload CSV with explicit dtype specified for mixed-type columns: `BuyerAgentAOR`, `ListAgentAOR`, `latfilled`, `lonfilled`, `BuyerAgencyCompensationType`, `OriginatingSystemName`, `OriginatingSystemSubName`, `ClosePrice`, `ListPrice`
5. Drop duplicate rows